### PR TITLE
AO/CO avoids recording in invalid hashtable for truncate replay offset=0

### PR DESCRIPTION
**Pull request by ashwinstar**
This commit is to fix PANIC on mirorr with "WAL contains references to
invalid pages" on xlog replay for truncate record with offset 0.

Primary creates the file first and then writes the xlog record for the
creation for AO tables similar to heap.  Hence, file can get created
on primary without writing xlog record if failure happens on primary
just after creating the file. This creates situation where VACUUM can
generate truncate record based on aoseg entry with eof 0 and file
present on primary. Then during replay mirror may not have the file,
as was never created on mirror. So, avoid adding the entry to invalid
hash table for truncate at offset zero (EOF=0).  This avoids mirror
PANIC, as anyways truncate to zero is same as file not present.

Co-authored-by: Alexandra Wang <lewang@pivotal.io>

This failure was seen in CI, super happy. Concourse inconsistent test execution time was able to trigger such a edge case and hit the panic on primary with all starts aligned.

## Here are some reminders before you submit the pull request
- [X] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [X] Pass `make installcheck`

### CODE REVIEW

File: src/test/isolation2/sql/uao_crash_compaction_column.sql [hunk 1]
```diff
@@ -155,3 +181,36 @@ SELECT gp_wait_until_triggered_fault('compaction_before_cleanup_phase', 1, 1);
...
+where c.role='p' and c.content=0), 'restart');
+4<:
+-- Shows entries for new files added to pg_aocsseg table. These are
+-- the entries next vacuum command will use to perform truncate.
+0U:SELECT segno,column_num,physical_segno,tupcount,modcount,state FROM gp_toolkit.__gp_aocsseg('crash_vacuum_in_appendonly_insert_1');
+-- generates truncate xlog record for all the files having entry in
+-- pg_aocsseg table.
+6:VACUUM crash_vacuum_in_appendonly_insert_1;
+-- Make sure mirror is able to successfully replay all the truncate
+-- record generated by doesn't encounter the "WAL contains references
```
Comment (asimrp) [inline src/test/isolation2/sql/uao_crash_compaction_column.sql]: typo: "... records generated and doesn't encounter ..."

File: src/test/isolation2/sql/uao_crash_compaction_column.sql [hunk 2]
```diff
@@ -5,6 +5,31 @@
...
+	i int; /* in func */
+	result bool; /* in func */
+begin /* in func */
+	i := 0; /* in func */
+	-- Wait until all mirrors has replayed up to flush location
+	loop /* in func */
+		SELECT flush_location = replay_location INTO result from gp_stat_replication where gp_segment_id = 0; /* in func */
+		if result then /* in func */
+			return true; /* in func */
+		end if; /* in func */
```
Comment (berlin-ab) [inline src/test/isolation2/sql/uao_crash_compaction_column.sql]: What does the /* in func */ comment do here?

Comment (ashwinstar) [reply src/test/isolation2/sql/uao_crash_compaction_column.sql]: Isolation2 frameworks grammar can't handle multi-line functions as one unit but instead treats them as separate SQL statements based on `;` character. Hence, adding these `/* in func */` comment is a hacky way to get functions defined in isolation2 framework and get treated as one unit. We need to enhance its parsing grammar to handle the functions.
